exchange/pki/models.py

```
@@ -284,7 +285,12 @@ def create_hostnameportsslconfig(self, url, pk):
         :param pk: The pk id of the ssl config
         :return: new HostnamePortSslConfig
         '''
-        ssl_config = SslConfig.objects.get(pk=pk)
+        # If the user tried to give a bad SslConfig,
+        # just assign it to the default
+        try:
+            ssl_config = SslConfig.objects.get(pk=pk)
+        except SslConfig.DoesNotExist:
+            ssl_config = SSL_DEFAULT_CONFIG
```

Oh. `SSL_DEFAULT_CONFIG` is [just a dict](https://github.com/boundlessgeo/exchange/blob/pki/exchange/pki/settings.py), not an `SslConfig` instance. We would need to create an `SslConfig` record, based upon `SSL_DEFAULT_CONFIG` and then use that. However, I'm not sure that is the right approach.

Review the following to see how `SSL_DEFAULT_CONFIG` is used now: https://github.com/boundlessgeo/exchange/blob/pki/exchange/pki/ssl_adapter.py#L178-L204

I think we _do not_ want a `HostnamePortSslConfig` created with a default SSL config, if this `try` block fails. Maybe the error should just be allowed to raise, then we log it and then let the registration fail (informing the user somehow)?

Side note: maybe we just change `SSL_DEFAULT_CONFIG` to _actually be_ an `SslConfig` instance, and add it to the `models` module, or leave it in `settings` (though I don't think `settings` should import `models`).

---

Hmmm, right that's a good point. So if this creation fails, it means somehow it was attempted to select a `pk` of an `SslConfig` that does not exist - so raising an error makes sense. I'm not sure how it reports to the user, though.

This should be a very hard thing to make happen in the first place, so I think as long as an error is raised it's fine.

---

Ok, thinking about this some more... This may fix a couple issues.

If we do have a `SSL_DEFAULT_CONFIG` that is actually an SslConfig instance, we could prepend that to the SslConfig `.all()` output and load that into the selector, maybe as 'Secure Default'. Then, there will _always_ be a default secure SSL config, ready to insert into the SslConfig model table, that a user can choose, *regardless of whether an admin has added any SslConfigs*. In other words, a default this is better than basic `requests` package SSL.

Again, this should still be optional for all `https://` registrations, where the user still has to _opt in_ by selecting even the default one. Probably the default for the widget should be `----------` or something similar, e.g. `---select----` (whatever the norm is there).

Then, later, if the SsLConfig associated with a hostname:port has for some reason been removed, the `SSL_DEFAULT_CONFIG` instance can be auto-saved to the SslModel with a name like 'Secure Default copy for <hostname:port>'. Basically, like your code above, but with auto-creation based upon the default.

Then, if an admin updates the system with new SslConfigs, they, or the user, can just 'realign' the existing registered service to a some new config. The user would do it on the registered service details page, while the admin can do it more efficiently via the admin console.

---

Adding the default to the selector widget, we should probably set it to `-1` for its `pk`, so the code above knows what to do if the user is not selecting an existing SslConfig, i.e. that a new default record needs auto-created.